mm2src/mm2_main/src/lp_swap/maker_swap.rs

```diff
@@ -1079,28 +1086,43 @@ impl MakerSwap {
             ]));
         }
 
-        let locktime = self.r().data.maker_payment_lock;
-        loop {
-            match self.maker_coin.can_refund_htlc(locktime).compat().await {
-                Ok(CanRefundHtlc::CanRefundNow) => break,
-                Ok(CanRefundHtlc::HaveToWait(to_sleep)) => Timer::sleep(to_sleep as f64).await,
-                Err(e) => {
-                    error!("Error {} on can_refund_htlc, retrying in 30 seconds", e);
-                    Timer::sleep(30.).await;
-                },
+        if self.taker_coin.can_be_released() {
+            let taker_payment = self.r().taker_payment.clone();
+            if let Some(payment) = taker_payment {
+                let fail_htlc_fut = self.taker_coin.fail_htlc_backwards(&payment.tx_hex);
+                if let Err(e) = fail_htlc_fut.compat().await {
+                    error!("Error {} on failing htlc backwards, the htlc will be failed back automatically when locktime expires!", e)
+                }
```

Comment:
In lightning, It's important to release the HTLCs as soon as possible so it can be used for inbound liquidity in other swaps, payments or routing. Since the maker knows the swap secret/preimage, they can release the taker payment as soon as the swap is known to have failed. On the other side, for `taker_swap`, if it's released early before refund, the maker has the preimage/secret and can spend the taker other coin payment, so the taker has to make sure his payment is refunded successfully first.

Reply:
The method used in `maker_swap` at the start of refund is called `on_start_maker_payment_refund` after fixing this https://github.com/KomodoPlatform/atomicDEX-API/pull/1592#discussion_r1063375403 https://github.com/KomodoPlatform/atomicDEX-API/blob/f1333688f3e75acae233eec9d6154a04a6b55d82/mm2src/coins/lightning.rs#L929
For symmetry, I can make the `MakerSwapOps`/`TakerSwapOps` traits as following
```rust
/// Maker specific swap operations
#[async_trait]
pub trait MakerSwapOps {
    /// Performs an action on maker payment on starting taker payment refund process if applicable
    async fn on_start_taker_payment_refund(&self, maker_payment: &[u8]) -> OnRefundResult<()>;
    /// Perform an action on maker payment on taker payment refund success if applicable
    async fn on_taker_payment_refund_success(&self, maker_payment: &[u8]) -> OnRefundResult<()>;
}

/// Taker specific swap operations
#[async_trait]
pub trait TakerSwapOps {
    /// Performs an action on taker payment on starting maker payment refund process if applicable
    async fn on_start_maker_payment_refund(&self, taker_payment: &[u8]) -> OnRefundResult<()>;
    /// Perform an action on taker payment on maker payment refund success if applicable
    async fn on_maker_payment_refund_success(&self, taker_payment: &[u8]) -> OnRefundResult<()>;
}
}
```
and call one method at the start of refund and the other on refund success, what do you think @artemii235?

Reply:
I see, thanks for clarification! `on_start/success` traits extension looks great, the only note is I would rename e.g. `on_start_taker_payment_refund` to `on_taker_payment_refund_start` for uniformity 🙂